### PR TITLE
fix: Misc fixes

### LINKED ISSUE #1283: bench new-app <app name>
Issue: Bug report

bench --version
5.7.5

**Describe the bug** :chart_with_downwards_trend:
Used an invalid name for new-app command. When prompted, input a valid name. But bench still remember the invalid name and use it to build new app.

**To Reproduce** :page_with_curl:
See screen capture.

Linux: ubuntu 20.04

### LINKED ISSUE #1292: Error installing version-13 "bench.exceptions.InvalidRemoteException: Frappe branch does not exist"
Issue: Easy Install

**Distro Information (Required)**

bench develop

**Command Run (Required)**

To start the installation, the following commands were executed:

sudo apt install python3-minimal build-essential python3-setuptools

export LC_ALL=C.UTF-8

wget https://raw.githubusercontent.com/frappe/bench/develop/install.py

Then I tried running this command:

sudo python3 install.py --production --site site.example --verbose

In another instance I tried this one:

sudo python3 install.py --version 13 --production --site site.example --verbose

Both produced the same error:

"bench.exceptions.InvalidRemoteException: Frappe branch does not exist"

Is there any particular reason for this to happen?

Reviewing the source code of the bench, I observe that the API:

https://api.github.com/repos/frappe/frappe/branches

The list that returns does not appear version-13 although from github you can see that branch.

The same happens with erpnext, the API does not return version-13

**Additional context**

Server Ubuntu 18.04

->>>>>>>>>>>>>>>Error<<<<<<<<<<<<<<-

fatal: [localhost]: FAILED! => {
    "changed": true,
    "cmd": [
        "bench",
        "init",
        "~frappe/frappe-bench",
        "--frappe-path",
        "https://github.com/frappe/frappe",
        "--frappe-branch",
        "version-13",
        "--python",
        "python3"
    ],
    "delta": "0:00:09.657527",
    "end": "2022-04-07 21:57:57.012289",
    "invocation": {
        "module_args": {
            "_raw_params": "bench init ~frappe/frappe-bench --frappe-path https://github.com/frappe/frappe --frappe-branch version-13 --python python3",
            "_uses_shell": false,
            "argv": null,
            "chdir": null,
            "creates": "/home/frappe/frappe-bench",
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": true
        }
    },
    "msg": "non-zero return code",
    "rc": 1,
    "start": "2022-04-07 21:57:47.354762",
    "stderr": "Aborted!",
    "stderr_lines": [
        "Aborted!"
    ],
    "stdout": "WARN: bench is installed in editable mode!\n\nThis is not the recommended mode of installation for production. Instead, install the package from PyPI with: `pip install frappe-bench`\n\nSetting Up Environment\n$ /usr/local/bin/virtualenv --quiet env -p python3\n$ /home/frappe/frappe-bench/env/bin/python -m pip install --quiet --upgrade pip\nTraceback (most recent call last):\n  File \"/home/frappe/.bench/bench/commands/make.py\", line 83, in init\n    verbose=verbose,\n  File \"/home/frappe/.bench/bench/utils/render.py\", line 95, in wrapper_fn\n    return fn(*args, **kwargs)\n  File \"/home/frappe/.bench/bench/utils/system.py\", line 78, in init\n    is_valid_frappe_branch(frappe_path=frappe_path, frappe_branch=frappe_branch)\n  File \"/home/frappe/.bench/bench/utils/__init__.py\", line 81, in is_valid_frappe_branch\n    raise InvalidRemoteException(\"Frappe branch does not exist\")\nbench.exceptions.InvalidRemoteException: Frappe branch does not exist\n\nERROR: There was a problem while creating /home/frappe/frappe-bench\nDo you want to rollback these changes? [y/N]: ",
    "stdout_lines": [
        "WARN: bench is installed in editable mode!",
        "",
        "This is not the recommended mode of installation for production. Instead, install the package from PyPI with: `pip install frappe-bench`",
        "",
        "Setting Up Environment",
        "$ /usr/local/bin/virtualenv --quiet env -p python3",
        "$ /home/frappe/frappe-bench/env/bin/python -m pip install --quiet --upgrade pip",
        "Traceback (most recent call last):",
        "  File \"/home/frappe/.bench/bench/commands/make.py\", line 83, in init",
        "    verbose=verbose,",
        "  File \"/home/frappe/.bench/bench/utils/render.py\", line 95, in wrapper_fn",
        "    return fn(*args, **kwargs)",
        "  File \"/home/frappe/.bench/bench/utils/system.py\", line 78, in init",
        "    is_valid_frappe_branch(frappe_path=frappe_path, frappe_branch=frappe_branch)",
        "  File \"/home/frappe/.bench/bench/utils/__init__.py\", line 81, in is_valid_frappe_branch",
        "    raise InvalidRemoteException(\"Frappe branch does not exist\")",
        "bench.exceptions.InvalidRemoteException: Frappe branch does not exist",
        "",
        "ERROR: There was a problem while creating /home/frappe/frappe-bench",
        "Do you want to rollback these changes? [y/N]: "
    ]
}

PLAY RECAP ******************************************************************************************************************************************************************************************
localhost                  : ok=71   changed=44   unreachable=0    failed=1    skipped=59   rescued=0    ignored=0

Traceback (most recent call last):
  File "install.py", line 497, in <module>
    install_bench(args)
  File "install.py", line 278, in install_bench
    run_playbook('site.yml', sudo=True, extra_vars=extra_vars)
  File "install.py", line 413, in run_playbook
    success = subprocess.check_call(args, cwd=playbooks_folder, stdout=log_stream, stderr=sys.stderr)
  File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ansible-playbook', '-c', 'local', 'site.yml', '-vvvv', '-e', '@/tmp/extra_vars.json', '--become', '--become-user=frappe']' returned non-zero exit status 2.

**Possible Solution**

The API returns within the list of branches the version-13 of frappe and erpnext

### LINKED ISSUE #1298: Github API pagination  ERROR
https://github.com/frappe/bench/blob/eb7e006bc4c9f946e37c8ca9598dd1ac1265bf98/bench/utils/__init__.py#L80

git_api_req = f"https://api.github.com/repos/{owner}/{repo}/branches"
github returns branches list in  pagination (10 per page) this calls only first page so can't validate branches prior to v13-xxx

### PR DESCRIPTION
**Pull request by phot0n**
Changes:
- use git's `ls-remote` command for validating frappe branch
- don't allow creating apps with names which start with int and/or have `.` in them

closes: https://github.com/frappe/bench/issues/1292, 
closes: https://github.com/frappe/bench/issues/1283, 
closes: https://github.com/frappe/frappe/issues/16524, 
closes: https://github.com/frappe/bench/issues/1298

### CODE REVIEW

**Comment by phot0n** [comment]
@Aradhya-Tripathi any reason this is needed over here? the function is very restrictive as it handles only github - so it will hinder anyone who tries to run `init` command if they have their frappe fork in gitlab or any other git server.
also, any particular reason why we're validating branches?

https://github.com/frappe/bench/blob/30d472dc6ba87635936d49f5d00767b619e76ba1/bench/utils/system.py#L78

**Comment by Aradhya-Tripathi** [comment]
> @Aradhya-Tripathi any reason this is needed over here? the function is very restrictive as it handles only github - so it will hinder anyone who tries to run `init` command if they have their frappe fork in gitlab or any other git server. also, any particular reason why we're validating branches?
> 
> https://github.com/frappe/bench/blob/30d472dc6ba87635936d49f5d00767b619e76ba1/bench/utils/system.py#L78

we can make this more dynamic for respective git servers but branch validation is required as it serves as an early fail system , during resolution if an app down the line as an invalid branch (branch that does not exist) it will fail (during resolution) before any of the dependencies are cloned.

**Comment by revant** [comment]
```python
import subprocess


r = subprocess.check_output(
    (
        "git",
        "ls-remote",
        "--heads",
        "https://github.com/frappe/frappe.git",
        "version-13",
    ),
    encoding="UTF-8",
)

if r:
    print("Branch exists")
```

more: https://stackoverflow.com/a/30524983

**Comment by phot0n** [comment]
Thanks @revant, we were going to implement it using `ls-remote` itself 😅 - probably @Aradhya-Tripathi has some College stuff going on so he wasn't able to get on it.

**Comment by revant** [comment]
```
frappe@a93cf37fb516:/workspace/development$ bench init --skip-redis-config-generation --frappe-branch version-13 frappe-delete
Setting Up Environment
$ /home/frappe/.local/bin/virtualenv --quiet env -p python3
$ /workspace/development/frappe-delete/env/bin/python -m pip install --quiet --upgrade pip
Traceback (most recent call last):
  File "/home/frappe/.bench/bench/commands/make.py", line 70, in init
    init(
  File "/home/frappe/.bench/bench/utils/render.py", line 95, in wrapper_fn
    return fn(*args, **kwargs)
  File "/home/frappe/.bench/bench/utils/system.py", line 78, in init
    is_valid_frappe_branch(frappe_path=frappe_path, frappe_branch=frappe_branch)
  File "/home/frappe/.bench/bench/utils/__init__.py", line 81, in is_valid_frappe_branch
    raise InvalidRemoteException("Frappe branch does not exist")
bench.exceptions.InvalidRemoteException: Frappe branch does not exist
```

Got this error recently on bench, Gavin referred me to this PR.

**Comment by phot0n** [comment]
yeah that's happening due to the api response getting paginated. this pr fixes it but we wanted to make it *not* github specific - and `ls-remote` works well in that case.

File: bench/utils/__init__.py [hunk 1]
```diff
@@ -53,32 +53,33 @@ def is_frappe_app(directory: str) -> bool:
...
+					"--heads",
+					frappe_path,
+					frappe_branch,
+				),
+				encoding="UTF-8",
+			)
+			if not ret:
+				raise InvalidRemoteException(message)
+		except subprocess.CalledProcessError:
+			raise InvalidRemoteException(message)
```
Comment (phot0n) [inline bench/utils/__init__.py]: the message here should be different?

We don't get any output when the git url is correct and branch is incorrect/does not exist,
though we do get an error when the git url is wrong - so the message over here should reflect the same? 
maybe something like: "frappe path/git url is invalid"

Screenshot 2022-04-23 at 5 57 06 PM

Comment (Aradhya-Tripathi) [reply bench/utils/__init__.py]: implemented by [bb10914](https://github.com/frappe/bench/pull/1294/commits/bb10914c435b86d65c1ee8080a70d9fcb3af7f3a)

**Pull request merged**